### ISSUE TITLE
[QA] Find mixed types for that psalm does not throw an error

### ISSUE THREAD

**Issue by mstilkerich** [vimeo/psalm#6322]
Hello,

  for one of my projects, psalm reports a type coverage of 99.8112% with errorLevel=1 (full config below). Now I am interested where the gaps are. I tried the ``--stats`` switch, and psalm will show me that there are still mixed types in some of the files.
How can I find them? My understanding was that psalm would issue an error in such cases with errorLevel=1.

### Psalm invocation
```
vendor/bin/psalm --no-cache --shepherd --report=testreports/psalm.txt --report-show-info=true --no-progress --stats
------------------------------
No errors found!
------------------------------

Checks took 8.33 seconds and used 166.784MB of memory
Psalm was able to infer types for 99.8112% of the codebase
-----------------
100.000% carddav.php (0 mixed)
100.000% dbmigrations/0007-replaceurlplaceholders/migrate.php (0 mixed)
100.000% dbmigrations/0017-accountentities2/migrate.php (0 mixed)
100.000% src/Addressbook.php (0 mixed)
97.500% src/Config.php (2 mixed)
... more here ...
```

### Full config
```xml
<?xml version="1.0"?>
<psalm
    errorLevel="1"
    resolveFromConfigFile="true"
    findUnusedVariablesAndParams="true"
    findUnusedPsalmSuppress="true"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xmlns="https://getpsalm.org/schema/config"
    xsi:schemaLocation="https://getpsalm.org/schema/config vendor/vimeo/psalm/config.xsd"
    >
<!--
    checkForThrowsInGlobalScope="true"
    findUnusedCode="true"
-->
    <projectFiles>
        <directory name="src/"/>
        <directory name="dbmigrations/"/>
        <file name="carddav.php"/>

        <directory name="tests/"/>
        <ignoreFiles>
            <directory name="tests/dbinterop/data" />
            <directory name="tests/unit/data" />
        </ignoreFiles>
    </projectFiles>

    <extraFiles>
        <directory name="roundcubemail/program/"/>
    </extraFiles>
    <plugins>
        <pluginClass class="Psalm\PhpUnitPlugin\Plugin"/>
    </plugins>

    <issueHandlers>
        <UnusedVariable>
            <errorLevel type="suppress">
                <directory name="tests/dbinterop/data" />
            </errorLevel>
        </UnusedVariable>
    </issueHandlers>
</psalm>
```

**Comment by orklah** [comment]
It may be because some mixed types are explicitly typed as such. I think I remember psalm don't emit issues when it happens but I'm not sure. Maybe share your config file so we can take a look?

**Comment by mstilkerich** [comment]
Hello,

 thanks for your reply. 

My config file is included above.

One of the source files with mixed types is this:
https://github.com/mstilkerich/rcmcarddav/blob/new_ui/src/Config.php

It does not contain explicit mixed types. Actually typing is pretty explicit from my point of view in that file.

**Comment by weirdan** [comment]
Applying the following patch brings the number of mixed to zero:
```patch
diff --git a/src/Config.php b/src/Config.php
index 7e362bc..841060f 100644
--- a/src/Config.php
+++ b/src/Config.php
@@ -147,7 +147,9 @@ class Config
     {
         if ($password == "%b") {
             if (
-                isset($_SESSION['oauth_token']['access_token'])
+                isset($_SESSION['oauth_token'])
+                && is_array($_SESSION['oauth_token'])
+                && isset($_SESSION['oauth_token']['access_token'])
                 && is_string($_SESSION['oauth_token']['access_token'])
             ) {
                 $credentials = [ 'bearertoken' => $_SESSION['oauth_token']['access_token'] ];
```

**Comment by orklah** [comment]
Seems like Psalm missed MixedArrayAccess then

**Comment by weirdan** [comment]
I wouldn't expect it emitted inside `isset` call. What Psalm could do, is to refine `$_SESSION['oauth_token']` to `array|ArrayAccess`.

**Comment by mstilkerich** [comment]
Hello,

 thanks for the hint. So my take is it requires more in-depths understanding of the inner workings of psalm to find these issues, there is no option for psalm to tell the locations. Concerning the concrete example above, I will probably be able to find quite a few of them, since `isset()` on a multi-level data structure is something I use quite frequently to make sure all levels exist. I'm not sure whether I want to modify the code as suggested though, as it seems detrimental to readability.
